### PR TITLE
Use semver package for semantic version comparison and precedence checking

### LINKED ISSUE #424: CLI version sorting potentially wrong
The extension obtains a list of CLI versions in order to find the most recent one. 

Currently, `versionCompare` treats `v1.2.3` and `v1.2.3-prerelease` as equal (by returning 0, because the `prerelease` component of `v1.2.3` is `undefined`) and the surrounding comparison function defers to creation date to disambiguate them. This is probably fine (since prereleases are likely to be released chronologically before the releases they correspond to) but if we ever released a sequence of CLI versions such as

- `v1.2.3-zyxw`
- `v1.2.3`
- `v1.2.3-abcd`

this would mean the sort order is implementation-defined: for the comparison function would imply a cycle containing

- `v1.2.3-zyxw < v1.2.3` (since `v1.2.3-zyxw` was released chronologically first) 
- `v1.2.3 < v1.2.3-abcd` (since `v1.2.3` was released chronologically first) 
- `v1.2.3-abcd < v1.2.3-zyxw` (since `abcd` precedes `zyxw` lexicographically).

Consider using the standard `semvar` package to ensure a consistent sort order.

### PR DESCRIPTION
**Pull request by henrymercer**
This PR refactors the CLI distribution manager to use the [semver](https://www.npmjs.com/package/semver) package.  This:

- fixes an issue with precedence checking where versions with prereleases were not compared correctly with versions without prereleases (#424)
- improves handling of cases where users have opted-in then opted-out of prereleases (see commit message of 1bc50f63b645c4dbf5ab0e875f6c51b9765930fd and the inline code comment for details)
- simplifies defining what versions of the CLI are compatible with the extension
- removes code implementing semver parsing and precedence checking

Other points to note:

- Semantic versions are parsed using `semver.parse` rather than `semver.coerce` since `semver.coerce` strips build metadata.
- Semantic versions are formatted using `version.raw` since this includes build metadata.
- Release creation dates are now compared using the `en-US` locale to remove potential locale dependence of version comparison.

## Checklist

- [x] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/master/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [x] Issues have been created for any UI or other user-facing changes made by this pull request.
- [x] `@github/product-docs-dsp` has been cc'd in all issues for UI or other user-facing changes made by this pull request.

Fixes #424

### CODE REVIEW

File: extensions/ql-vscode/src/distribution.ts [hunk 1]
```diff
@@ -100,10 +115,18 @@ export class DistributionManager implements DistributionProvider {
     return result.kind !== FindDistributionResultKind.NoDistribution;
   }
 
+  public async getCodeQlPathWithoutVersionCheck(): Promise<string | undefined> {
+    const distribution = await this.getDistributionWithoutVersionCheck();
+    if (distribution === undefined) {
+      return undefined;
+    }
+    return distribution.codeQlPath;
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/distribution.ts:123]: nit: 

```suggestion
    return distribution?.codeQlPath;
```

File: extensions/ql-vscode/src/distribution.ts [hunk 2]
```diff
@@ -150,9 +182,9 @@ export class DistributionManager implements DistributionProvider {
    */
   public async checkForUpdatesToExtensionManagedDistribution(
     minSecondsSinceLastUpdateCheck: number): Promise<DistributionUpdateCheckResult> {
-    const codeQlPath = await this.getCodeQlPathWithoutVersionCheck();
+    const distribution = await this.getDistributionWithoutVersionCheck();
     const extensionManagedCodeQlPath = await this._extensionSpecificDistributionManager.getCodeQlPathWithoutVersionCheck();
-    if (codeQlPath !== undefined && codeQlPath !== extensionManagedCodeQlPath) {
+    if (distribution !== undefined && distribution.codeQlPath !== extensionManagedCodeQlPath) {
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/distribution.ts:187]: nit:

```suggestion
    if (distribution?.codeQlPath !== extensionManagedCodeQlPath) {
```

File: extensions/ql-vscode/src/extension.ts [hunk 3]
```diff
@@ -182,11 +192,25 @@ export async function activate(ctx: ExtensionContext): Promise<void> {
...
-        helpers.showAndLogWarningMessage("The current version of the CodeQL CLI is incompatible with this extension.");
+      case FindDistributionResultKind.IncompatibleDistribution: {
+        const fixGuidanceMessage = (() => {
+          switch (result.distribution.kind) {
+            case DistributionKind.ExtensionManaged:
+              return "Please update the CodeQL CLI by running the \"CodeQL: Check for CLI Updates\" command.";
+            case DistributionKind.CustomPathConfig:
+              return `Please update the \"CodeQL CLI Executable Path\" setting to point to a CLI in the version range ${codeQlVersionRange}.`;
+            case DistributionKind.PathEnvironmentVariable:
+              return `Please update the CodeQL CLI on your PATH to a version in the range ${codeQlVersionRange}, or ` +
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/extension.ts:205]:
```suggestion
              return `Please update the CodeQL CLI on your PATH to a version compatible with ${codeQlVersionRange}, or ` +
```

File: extensions/ql-vscode/src/extension.ts [hunk 4]
```diff
@@ -182,11 +192,25 @@ export async function activate(ctx: ExtensionContext): Promise<void> {
...
+                `set the \"CodeQL CLI Executable Path\" setting to the path of a CLI in the version range ${codeQlVersionRange}.`;
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/extension.ts:206]:
```suggestion
                `set the \"CodeQL CLI Executable Path\" setting to the path of a CLI version compatible with ${codeQlVersionRange}.`;
```

**Comment by henrymercer** [comment]
Thanks @aeisenberg, I agree with all of your suggestions.

File: extensions/ql-vscode/src/vscode-tests/no-workspace/distribution.test.ts [hunk 5]
```diff
@@ -84,13 +85,30 @@ describe("Releases API consumer", () => {
...
 
-    const latestRelease = await consumer.getLatestRelease({
-      description: "2.*.*",
-      isVersionCompatible: version => version.majorVersion === 2
-    });
+    const latestRelease = await consumer.getLatestRelease(new semver.Range("2.*.*"));
     expect(latestRelease.id).to.equal(1);
   });
 
+  it("picking latest release: release passes additional compatibility test if additional compatibility test specified", async () => {
```
[the remark this answers is not in the document]
Comment (adityasharad) [reply extensions/ql-vscode/src/vscode-tests/no-workspace/distribution.test.ts:92]: Also want to be sure we have a test case that replicates the original problem - two releases that differ only in the prerelease version.

**Pull request merged**